src/editors/EditorHeader.jsx

```diff
@@ -1,44 +1,112 @@
 import {
-  ActionRow, IconButton, Icon, ModalDialog,
+  ActionRow, IconButton, Icon, ModalDialog, Form,
 } from '@edx/paragon';
-import PropTypes from 'prop-types';
-import React, { useContext } from 'react';
-import { Close } from '@edx/paragon/icons';
+import React, { useContext, useState, useEffect } from 'react';
+import { Edit, Close } from '@edx/paragon/icons';
 import EditorPageContext from './EditorPageContext';
-import { ActionStates, mapBlockTypeToName } from './data/constants';
+import { ActionStates } from './data/constants';
 
-const EditorHeader = ({ title }) => {
-  const { unitUrl, unitUrlLoading, studioEndpointUrl } = useContext(EditorPageContext);
+/* The Header component renders based on three states:
+    loading -> loaded but not editing -> editing. */
+
+const EditorHeader = () => {
+  const {
+    titleRef, editorRef, setBlockTitle,
+    blockValue, blockLoading,
+    unitUrl, unitUrlLoading, studioEndpointUrl,
+  } = useContext(EditorPageContext);
+
+  const [editing, setEditing] = useState(false);
+
+  const [title, setTitle] = useState('');
+  useEffect(() => {
+    if (blockLoading === ActionStates.FINISHED) {
+      setTitle(blockValue ? blockValue.data.display_name : '');
+    }
+  }, [blockLoading]);
+
+  const handleChange = (e) => {
+    setTitle(e.target.value);
+  };
+
+  const handleKeyDown = (e) => {
+    if (e.key === 'Enter') { setEditing(false); }
+    if (e.key === 'Tab' && editorRef) {
+      e.preventDefault();
+      editorRef.current.focus();
+    }
+  };
+
+  const handleOnBlur = () => {
+    setBlockTitle(title);
+    setEditing(false);
+  };
 
   const onCancelClicked = () => {
     if (unitUrlLoading === ActionStates.FINISHED) {
       const destination = `${studioEndpointUrl}/container/${unitUrl.data.ancestors[0].id}`;
       window.location.assign(destination);
     }
   };
+
+  let titleDisplay = 'Loading...';
```

Comment:
Why are we sometimes rendering a string, and sometimes rendering a react component here, depending on the loading state? I would caution that might not be great for the internationalization of the product. That can be easily fixed by replacing strings with formatted messages.

 Also, I'm not super comfortable with the  using a let here. It can get out of sync with the render in odd ways. I'll add a suggestion down below as to how we might do this behavior a little more cleanly.